apps/heft/src/plugins/TypeScriptPlugin/Eslint.ts

```diff
@@ -0,0 +1,111 @@
+
+import * as path from 'path';
+import * as crypto from 'crypto';
+import * as TEslint from 'eslint';
+
+import { LinterBase, ILinterBaseOptions } from './LinterBase';
+import { IExtendedSourceFile } from './internalTypings/TypeScriptInternals';
+import { IColorableSequence, Colors } from '@rushstack/node-core-library';
+
+interface IEslintOptions extends ILinterBaseOptions {
+  eslintPackagePath: string;
+}
+
+export class Eslint extends LinterBase<TEslint.ESLint.LintResult> {
+  private readonly _eslintPackage: typeof TEslint;
+
+  private _eslintCli: TEslint.CLIEngine;
+  private _eslint: TEslint.ESLint;
+  private _eslintBaseConfiguration: any; // eslint-disable-line @typescript-eslint/no-explicit-any
+  private _lintResult: TEslint.ESLint.LintResult[];
+
+  public constructor(options: IEslintOptions) {
+    super('eslint', options);
+
+    this._eslintPackage = require(options.eslintPackagePath);
+  }
+
+  public printVersionHeader(): void {
+    this._terminal.writeLine(`Using ESLint version ${this._eslintPackage.Linter.version}`);
+  }
+
+  public reportFailures(): void {
+    const eslintFailureLogMessages: (string | IColorableSequence)[][] = [];
+
+    for (const eslintFileResult of this._lintResult) {
+      const buildFolderRelativeFilePath: string = path.relative(
+        this._buildFolderPath,
+        eslintFileResult.filePath
+      );
+      for (const message of eslintFileResult.messages) {
+        // https://eslint.org/docs/developer-guide/nodejs-api#◆-lintmessage-type
+        const severity: string = message.severity === 1 ? 'WARNING' : 'ERROR';
+        eslintFailureLogMessages.push([
+          '  ',
+          Colors.yellow(`${severity}: ${buildFolderRelativeFilePath}:${message.line}:${message.column}`),
+          ' - ',
+          Colors.yellow(message.ruleId ? `(${message.ruleId}) ${message.message}` : message.message)
+        ]);
+      }
+    }
+
+    if (eslintFailureLogMessages.length > 0) {
+      this._terminal.writeWarningLine(
+        `Encountered ${eslintFailureLogMessages.length} ESLint error${
+          eslintFailureLogMessages.length > 1 ? 's' : ''
+        }:`
+      );
+      for (const eslintFailureLogMessage of eslintFailureLogMessages) {
+        this._terminal.writeWarningLine(...eslintFailureLogMessage);
+      }
+    }
+  }
+
+  protected get cacheVersion(): string {
+    const eslintConfigHash: crypto.Hash = crypto
+      .createHash('sha1')
+      .update(JSON.stringify(this._eslintBaseConfiguration));
+    const eslintConfigVersion: string = `${this._eslintPackage.Linter.version}_${eslintConfigHash.digest(
+      'hex'
+    )}`;
+
+    return eslintConfigVersion;
+  }
+
+  protected async initializeAsync(): Promise<void> {
+    this._eslint = new this._eslintPackage.ESLint({
+      cwd: this._buildFolderPath
+    });
+
+    this._eslintBaseConfiguration = await this._eslint.calculateConfigForFile(this._linterConfigFilePath);
+
+    this._eslintCli = new this._eslintPackage.CLIEngine({
```

Comment:
ESLint goes all nuts with `cosmiconfig` alternative filenames: `.eslintrc.js`, `.eslintrc.cjs`, `.eslintrc.yaml`, `.eslintrc.yml`, `.eslintrc.json`, `.eslintrc`, `package.json` (in that order of precedence).

Since Heft is opinionated, IMO it should reject all names other than `.eslintrc.js` for a couple reasons:
- In a monorepo, it's a confusing experience to have to consider 7 different filenames when searching for these files.
- In practice `.js` is the only generally useful file extension, because due to historical reasons basic config problems cannot be solved without scripting (e.g. [__dirname](https://github.com/microsoft/rushstack/blob/8b0a39576caf5a1cfb34321543fe118f17196067/libraries/stream-collator/.eslintrc.js#L6) in our setup)